### ISSUE TITLE
Some geometry appearing in the wrong place

### ISSUE THREAD

**Issue by miccTronic** [Five-Damned-Dollarz/io_scene_jupex#8]
Hi Amphos,
thanks a lot for your great work on the Jupiter EX importer, it was really useful in learning more about the maps from FEAR!

I have one issue though: About 80% of the geometry imports properly, but around 20% is somehow "stuck" in a section around the map origin (0, 0, 0). It seems that it's mostly the dynamic objects, but also some sprites like graffiti. I checked the face/vertex coordinates in the .world00p file, they are really given like that, i.e. placed in the origin instead of at their proper location.

I also noticed that for some of those, only a single instance appears in the maps (both in the BSP and the RenderSurfaces section), although there should be many scattered around the level. For example, there is just 1 paint_can (per color) in the map, whereas they appear all over the place in shelves (in the "Intro" level). I would assume that there is some structure that "places" these mesh instances somewhere, but I couldn't find it.
In fact I spent the last couple of days searching the world file up and down with the help of your .bt template (thanks also for that!), but I wasn't successfull looking e.g. in the render tree or sectors. The only place I did find the individual instances is in the "Objects" section, where they are usually of type "WorldModel" with their proper position and rotation. BUT here the thing is that this WorldModel contains no link to a model, material or RenderSurface, so I'm kind of stuck again.

Since you appear to be the only expert around on Jupiter EX, I was hoping you might have encountered the issue too, or maybe have an idea how the heck the world files store the proper positioning info. :-)
Thanks a lot in advance!

**Comment by Five-Damned-Dollarz** [comment]
Centering dynamic objects around 0,0,0 is necessary for correct transforms when assigned to a dynamic object instance. The plugin supports *limited* dynamic objects if you tick Import Objects; it will create empties at the correct positions and set them as parent to the relevant meshes to move them there.
I haven't gotten around to the next step of that where it would instance a shared mesh to multiple empties. You can see the relevant code in WorldObjects.py, but it's extremely bare at the moment.

**Comment by miccTronic** [comment]
The WorldObjects part I found, but ok now I see there can be multiple names per WorldModel/BSP, I just overwrote the name each time, so each BSP had just one name assigned. :-/

But still, now I have instances of the WorldModels, but not the RenderMeshes, the WorldModels just being collision geometry, you don't want to draw those. Do you by any chance know how to get from a BSP to the RenderMesh so I know which materials to use at those instance positions? The RenderMeshes don't seem to have names...

**Comment by Five-Damned-Dollarz** [comment]
They're in sequential order and should be 1:1

**Comment by miccTronic** [comment]
Hm BSP and render faces count are quite different, but I think I'm getting the idea... I need to look at the RenderTree structure, right? Thanks, that helped a lot!! 😊

**Comment by Five-Damned-Dollarz** [comment]
Yeah, there's still a lot left to solve about the render surfaces (even worse of in the Loki variant). I'm not focused on this right now, but if you figure anything out be sure to share it

**Comment by miccTronic** [comment]
I will :-) I've been working my way up from LT1, the objects haven't changed so much, but render structures did. Regarding sectors, there's a good article on lighting in the SDK docs (public tools v2) that explains a bit.
Will probably also look into Loki and possibly Yeti and some point.
